### PR TITLE
added VDE functionality

### PR DESCRIPTION
**Pull request by MSharq**
Added VDE functionality, now one can use internet with vde-switch. This could be a useful enhancement, if somebody wants to use internet in macemu with vde-switch.

### CODE REVIEW

File: BasiliskII/src/Unix/main_unix.cpp [hunk 1]
```diff
@@ -436,6 +439,17 @@ int main(int argc, char **argv)
...
+		} else if (strcmp(argv[i], "--switch") == 0) {
+			argv[i] = NULL;
+			if (argv[++i] == NULL) {
+				printf("switch address not defined\n");
+				usage(argv[0]);
+			}
+			int vde_str_len = strlen(argv[i]);
+			vde_sock = (char*)malloc(vde_str_len);
+			memset(vde_sock,NULL,vde_str_len);
+			strncpy(vde_sock,argv[i],vde_str_len);
```
Comment (asvitkine) [inline BasiliskII/src/Unix/main_unix.cpp]: I think lines 448 to 451 can be replaced with strdup(argv[i]).

Although, do we really need to copy it? If we're setting to NULL in the array, why not just point to the original content?

File: BasiliskII/src/Unix/ether_unix.cpp [hunk 2]
```diff
@@ -115,6 +119,7 @@ static pthread_t slirp_thread;				// Slirp reception thread
 static bool slirp_thread_active = false;	// Flag: Slirp reception threadinstalled
 static int slirp_output_fd = -1;			// fd of slirp output pipe
 static int slirp_input_fds[2] = { -1, -1 };	// fds of slirp input pipe
+VDECONN *vde_conn;
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: I will do this. Thanks!

File: BasiliskII/src/Unix/ether_unix.cpp [hunk 3]
```diff
@@ -925,13 +988,17 @@ void ether_do_interrupt(void)
 		} else
 #endif
 		{
+			if (net_if_type == NET_IF_VDE) {
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: i do not understand "fix indent on lines 995 and later", they are already indented. Probably I am not understanding your expectation

File: BasiliskII/src/Unix/ether_unix.cpp [hunk 4]
```diff
@@ -749,9 +794,27 @@ static int16 ether_do_write(uint32 arg)
...
 #endif
-	if (write(fd, packet, len) < 0) {
+	if (net_if_type == NET_IF_VDE) {
+		if (fd == -1) {	// which means vde service is not running
+			D(bug("WARNING: Couldn't transmit VDE packet\n"));
+			return excessCollsns;
+		}
+
+		if (vde_conn == NULL)
+		{
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: done

File: BasiliskII/src/Unix/ether_unix.cpp [hunk 5]
```diff
@@ -749,9 +794,27 @@ static int16 ether_do_write(uint32 arg)
...
+			D(bug("WARNING: vde_conn is NULL\n"));
+			return -1;
+		}
+
+		do {
+			len = vde_send(vde_conn,packet,sizeof(packet),0);
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: done

File: BasiliskII/src/Unix/main_unix.cpp [hunk 6]
```diff
@@ -374,7 +376,8 @@ static void usage(const char *prg_name)
 		"  --display STRING\n    X display to use\n"
 		"  --break ADDRESS\n    set ROM breakpoint in hexadecimal\n"
 		"  --loadbreak FILE\n    load breakpoint from FILE\n"
-		"  --rominfo\n    dump ROM information\n", prg_name
+		"  --rominfo\n    dump ROM information\n"
+		"  --switch SWITCH_PATH\n    vde switch address\n", prg_name
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/main_unix.cpp]: changed vde switch to vde_switch.

File: BasiliskII/src/Unix/ether_unix.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: do you mean to indent these "#if defined(__linux__)" and others.. actually in the code I saw that #if or #else directives always start from the start of the line, so I did the same here.

File: BasiliskII/src/Unix/ether_unix.cpp [hunk 7]
```diff
@@ -250,6 +255,10 @@ bool ether_init(void)
 	else if (strcmp(name, "slirp") == 0)
 		net_if_type = NET_IF_SLIRP;
 #endif
+	else if (strcmp(name, "vde") == 0){
+		printf("selected Ethernet device type VDE\n");
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: add printfs for all the others

File: BasiliskII/src/Unix/ether_unix.cpp [hunk 8]
```diff
@@ -749,9 +794,27 @@ static int16 ether_do_write(uint32 arg)
 		write(slirp_input_fd, &len, sizeof(len));
 		write(slirp_input_fd, packet, len);
 		return noErr;
-	} else
```
[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: Yes, you are right. That was my mistake.

File: BasiliskII/src/Unix/ether_unix.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (asvitkine) [reply BasiliskII/src/Unix/ether_unix.cpp]: I meant that lines1004 and 1006 aren't indented properly as a result of your else on line 1000.

But one thing you can do is make line 995 an "else if (net_if_type == NET_IF_VDE) {" and then indent doesn't have to change on those other lines.

[the remark this answers is not in the document]
Comment (MSharq) [reply BasiliskII/src/Unix/ether_unix.cpp]: I got it, sorry I didn't understand at first, I have indented 1004 and 1006.

line 995 cannot be converted to "else if(net_if_type == NET_IF_VDE) {"  it has to stay "else", as the code I have added goes for "VDE", but the code already present goes for "sheep_net device". After passing else condition, we have two options one VDE and the other sheep_net.

File: BasiliskII/src/Unix/main_unix.cpp (hunk 1 above)
Comment (asvitkine) [reply BasiliskII/src/Unix/main_unix.cpp]: Ping

Comment (asvitkine) [reply BasiliskII/src/Unix/main_unix.cpp]: I'm not following. You are passing pointers to the vde_open() API...

Comment (MSharq) [reply BasiliskII/src/Unix/main_unix.cpp]: again it was my mistake. Yes, I am passing pointers to vde_open, you were right, I have changed the code, there is no need for any malloc or copy, just a pointer passing will do.

**Pull request merged**